src/trace/harTracer.ts

```diff
@@ -0,0 +1,298 @@
+  private _lastPage = 0;
+  private _barrierPromises = new Map<Promise<void>, Page>();
+
+  constructor(context: BrowserContext, options: HarOptions) {
+    this._browserName = context._browser._options.name;
+    this._options = options;
+    this._log = {
+      version: '1.2',
+      creator: {
+        name: 'Playwright',
+        version: require('../../package.json')['version'],
+      },
+      browser: {
+        name: context._browser._options.name,
+        version: context._browser.version()
+      },
+      pages: [],
+      entries: []
+    };
+    context.on(BrowserContext.Events.Page, page => this._onPage(page));
+  }
+
+  private _onPage(page: Page) {
+    const pageEntry: har.Page = {
+      startedDateTime: new Date(),
+      id: `page_${this._lastPage++}`,
+      title: '',
+      pageTimings: {
+        onContentLoad: -1,
+        onLoad: -1,
+      },
+    };
+    this._pageEntries.set(page, pageEntry);
+    this._log.pages.push(pageEntry);
+    page.on(Page.Events.Request, (request: network.Request) => this._onRequest(page, request));
+    page.on(Page.Events.Response, (response: network.Response) => this._onResponse(page, response));
+
+    page.on(Page.Events.DOMContentLoaded, () => {
+      const promise = page.mainFrame()._evaluateExpression(String(() => {
+        return {
+          title: document.title,
+          domContentLoaded: performance.timing.domContentLoadedEventStart,
+        };
+      }), true, undefined).then(result => {
+        pageEntry.title = result.title;
+        pageEntry.pageTimings.onContentLoad = result.domContentLoaded;
+      }).catch(() => {});
+      this._barrierPromises.set(promise, page);
+    });
+    page.on(Page.Events.Load, () => {
+      const promise = page.mainFrame()._evaluateExpression(String(() => {
+        return {
+          loaded: performance.timing.loadEventStart,
+        };
+      }), true, undefined).then(result => {
+        pageEntry.pageTimings.onLoad = result.loaded;
+      }).catch(() => {});
+      this._barrierPromises.set(promise, page);
+    });
+    page.on(Page.Events.Close, () => {
+      for (const [promise, p] of this._barrierPromises) {
+        if (page === p)
+          this._barrierPromises.delete(promise);
+      }
+    });
+  }
+
+  private _onRequest(page: Page, request: network.Request) {
+    const pageEntry = this._pageEntries.get(page)!;
+    const url = new URL(request.url());
+
+    const harEntry: har.Entry = {
+      pageref: pageEntry.id,
+      startedDateTime: new Date(),
+      time: -1,
+      request: {
+        method: request.method(),
+        url: request.url(),
+        httpVersion: 'HTTP/1.1',
+        cookies: [],
+        headers: [],
+        queryString: [...url.searchParams].map(e => ({ name: e[0], value: e[1] })),
+        postData: undefined,
+        headersSize: -1,
+        bodySize: -1,
+      },
+      response: {
+        status: -1,
+        statusText: '',
+        httpVersion: 'HTTP/1.1',
```

Comment:
What about http v2, quick or v3 based traffic?

Reply:
I don't think it requires documenting, I'd rather wait for reports where users would want to see h2 or quic there. Har predates both and its timing model doesn't really make much sense with them.